File: engine/core/painting/Point.cpp
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "sky/engine/config.h"
#include "sky/engine/core/painting/Point.h"

#include "sky/engine/core/script/dom_dart_state.h"
#include "sky/engine/tonic/dart_error.h"
#include "base/logging.h"

namespace blink {

// Convert dartPoint.x,y ==> SkPoint.
Point DartConverter<Point, void>::FromArgumentsWithNullCheck(
    Dart_NativeArguments args,
    int index,
    Dart_Handle& exception) {
  Point result;
  result.is_null = true;

  Dart_Handle dartPoint = Dart_GetNativeArgument(args, index);
  DCHECK(!LogIfError(dartPoint));

  Dart_Handle x_value =
      Dart_GetField(dartPoint, DOMDartState::Current()->x_handle());
  Dart_Handle y_value =
      Dart_GetField(dartPoint, DOMDartState::Current()->y_handle());

  double x = 0.0, y = 0.0;
  Dart_Handle err = Dart_DoubleValue(x_value, &x);
  DCHECK(!LogIfError(err));
  err = Dart_DoubleValue(y_value, &y);
  DCHECK(!LogIfError(err));
  result.sk_point.set(x, y);
  result.is_null = false;
  return result;
}

} // namespace blink

File: engine/core/painting/Point.dart
// Copyright 2015 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

part of dart.sky;

/// Holds 2 floating-point coordinates.
class Point {
  const Point(this.x, this.y);

  final double x;
  final double y;

  static const Point origin = const Point(0.0, 0.0);

  bool operator ==(other) => other is Point && x == other.x && y == other.y;
  Point operator -() => new Point(-x, -y);
  Offset operator -(Point other) => new Offset(x - other.x, y - other.y);
  Point operator +(Offset other) => new Point(x + other.dx, y + other.dy);
  Rect operator &(Size other) => new Rect.fromLTWH(x, y, other.width, other.height);

  // does the equivalent of "return this - Point.origin"
  Offset toOffset() => new Offset(x, y);

  int get hashCode {
    int result = 373;
    result = 37 * result + x.hashCode;
    result = 37 * result + y.hashCode;
    return result;
  }
  String toString() => "Point($x, $y)";
}
